kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+                                 key)
+                    return False
+
+                # Check if the first items in the list match
+                if not match_item(pattern_value[0], kcidb_obj[key][0]):
+                    return False
+            else:
+                # Match individual items
+                if not match_item(pattern_value, kcidb_obj[key]):
+                    return False
+
+        if 'tests' in pattern_obj and 'log_regex' in pattern_obj['tests'][0]:
+            return self.check_log_match(kcidb_obj, pattern_obj, "tests")
+        if 'builds' in pattern_obj and 'log_regex' in pattern_obj['builds'][0]:
+            return self.check_log_match(kcidb_obj, pattern_obj, "builds")
+
+        return True
+
+    def check_log_match(self, kcidb_io_object, pattern_object, key):
+        def search_regex(log_regex, log_excerpt, log_url):
+            if re.search(log_regex, log_excerpt):
+                return True
+
+            log_content = get_log(log_url)
+            if log_content and re.search(log_regex, log_content):
+                return True
+
+            return False
+
+        log_regex = pattern_object[key][0].get("log_regex")
+        if log_regex and key in kcidb_io_object and kcidb_io_object[key]:
+            log_excerpt = kcidb_io_object[key][0].get("log_excerpt", "")
+            log_url = kcidb_io_object[key][0].get("log_url", "")
+            return search_regex(log_regex, log_excerpt, log_url)
+        return False
+
+
+class IncidentGenerator:
+    """Class to generate incidents"""
+    def __init__(self, db_name=DB_NAME):
+        self.db = PatternDatabase(db_name)
+        self.validator = PatternValidator()
+
+    def create_incident(self, kcidb_io_object, issue_id, issue_version):
+        """Create and return an incident object"""
+        if 'tests' in kcidb_io_object and kcidb_io_object['tests']:
+            type_id_key = "test_id"
+            type_id_value = kcidb_io_object['tests'][0]['id']
+        elif 'builds' in kcidb_io_object and kcidb_io_object['builds']:
+            type_id_key = "build_id"
+            type_id_value = kcidb_io_object['builds'][0]['id']
+        else:
+            raise ValueError("The KCIDB IO object must contain at least "
+                             "one non-empty test or build")
+
+        unique_string = f"{issue_id}{issue_version}{type_id_value}"
+        incident_id = f"{ORIGIN}:{hashlib.sha256(unique_string.encode(
+                      )).hexdigest()}"
+
+        return {
+            'id': incident_id,
+            'origin': ORIGIN,
+            'issue_id': issue_id,
+            'issue_version': issue_version,
+            'present': True,
+            type_id_key: type_id_value,
+        }
+
+    def generate_incident_on_match(self, kcidb_io_object, issue_id,
+                                   issue_version, issue_pattern_object):
+        """Generate incident if issue pattern is found in a build/test
+        object"""
+        incidents = []
```

Comment:
it seems this is returning a single instance, we can remove the array. This was probably a left over of some of my experimentations